aws/resource_aws_cloudhsm2_cluster.go

```diff
@@ -0,0 +1,384 @@
+		},
+
+		Schema: map[string]*schema.Schema{
+			"backup_identifier": {
+				Type:     schema.TypeString,
+				Computed: false,
+				Optional: true,
+				ForceNew: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+
+			"hsm_type": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateCloudHsm2HsmType,
+			},
+
+			"subnet_ids": {
+				Type:     schema.TypeSet,
+				Required: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+				ForceNew: true,
+			},
+
+			"cluster_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"vpc_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"cluster_certificates": {
+				Type:     schema.TypeSet,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"cluster_certificate": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"cluster_csr": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"aws_hardware_certificate": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"hsm_certificate": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"manufacturer_hardware_certificate": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+
+			"security_group_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"cluster_state": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceAwsCloudHsm2ClusterImport(
+	d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	d.Set("cluster_id", d.Id())
+	return []*schema.ResourceData{d}, nil
+}
+
+func describeCloudHsm2Cluster(clusterId string, meta interface{}) (*cloudhsmv2.Cluster, error) {
+	conn := meta.(*AWSClient).cloudhsmv2conn
+
+	filters := []*string{&clusterId}
+	result := int64(1)
+	out, err := conn.DescribeClusters(&cloudhsmv2.DescribeClustersInput{
+		Filters: map[string][]*string{
+			"clusterIds": filters,
+		},
+		MaxResults: &result,
+	})
+	if err != nil {
+		log.Printf("[WARN] Error on retrieving CloudHSMv2 Cluster (%s) when waiting: %s", clusterId, err)
+		return nil, err
+	}
+
+	var cluster *cloudhsmv2.Cluster
+
+	for _, c := range out.Clusters {
+		if *c.ClusterId == clusterId {
+			cluster = c
+		}
+	}
+	return cluster, nil
+}
+
+func resourceAwsCloudHsm2ClusterRefreshFunc(
+	d *schema.ResourceData, meta interface{}) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		cluster, err := describeCloudHsm2Cluster(d.Id(), meta)
+
+		if cluster == nil {
+			return 42, "destroyed", nil
+		}
+
+		if cluster.State != nil {
+			log.Printf("[DEBUG] CloudHSMv2 Cluster status (%s): %s", d.Id(), *cluster.State)
+		}
+
+		return cluster, *cluster.State, err
```

Comment:
To prevent potential `nil` panics, we should use the SDK provided helper function here: `return cluster, aws.StringValue(cluster.State), err`